### ISSUE TITLE
About the number of samples at each subEpoch

### ISSUE THREAD

**Issue by YilinLiu97** [josedolz/LiviaNET#18]
Hi @josedolz ,  I'm wondering, would 1000 be too much for the number of samples? E.g. for a dataset of 5 subjects, 200 samples would be extracted from each subject in a subEpoch, but there would be many redundant samples among these 200 ones.  previously I got pretty good results on LiviaNET with my own dataset, and now I'm trying to replicate it on another framework using the exact same architecture and hyperparameters but with no success. I found that this parameter would critically influence the training (loss fluctuates greatly all the time and is hard to get decreased) but am wondering why this was not an issue when using the LiviaNET. 

Thanks.

**Comment by josedolz** [comment]
Hi @YilinLiu97 

It is true that when sampling a 3D volume in smaller sub-volumes some information maybe repeated across different patches, but this does not mean samples are redundant between them. You can actually sample thousand of small sub-volumes from a larger volume and not repeat the same patch at once. For your question, it is difficult to answer with only that information, as they might be many other reasons. When you do the sampling, do you sample from any point of the volume, so that you allow all background (i.e., black) patches to be processed?

Jose.

**Comment by YilinLiu97** [comment]
Yes, but my volume is only 256x256x256, so there will surely be some redundant ones among 200 samples with patch size 27x27x27.  With too many redundant samples and a small batch size (say, 10), my observation is that the network tend to "memorize" each sample instead of learning to find a global minima for all subjects, which kindof explains why the loss always fluctuates and hard to get decreased. 

And yes I sampled from any point of the volume. Specifically, the # of samples from each class are roughly equal. 

I'm still trying to replicate the result. Now the loss is pretty low but the seg result is not getting better and actually, quite bad. 
![screenshot 2018-05-28 12 10 08](https://user-images.githubusercontent.com/26695815/40629088-b1e2f98c-628e-11e8-8985-0e9f735f4d1e.png)

**Comment by josedolz** [comment]
Hi @YilinLiu97 

let's see. If you do not allow overlapping between samples (i.e. sub patches of size 27x27x27) you should be able to get around 700 samples per image on the whole volume. If you constraint the sample selection to a ROI where at least a class (non-background) should be always there, you can go to 300 maybe. So even in this case there are no sub-patches repeated. So I do not see what you mean with redundant information. The think is that you should allow overlapping between samples. In that way you can get millions of samples without having exactly the same information (voxel-wise). In addition, just in case you do it, I wouldn't include samples with all the voxels in black (that is why we employ a generated ROI during training). Another thing, just be sure to shuffle all the samples once you get them from all the volumes. 

I hope that helps.

Jose.

**Comment by YilinLiu97** [comment]
Hi @josedolz , thank you for the detailed response. I'm implementing your architecture on another framework, so I'm actually not totally in charge of the sampling process. For the sampler there are several options there: balanced and random. I've tried both but the results were both as bad as the one above, though the loss is quite low. The samples were shuffled. 

I noticed that your sampling strategy seems to attach equal weight to both the foreground and the background (?), and it indeed worked very well on my dataset. Interestingly, when I tried another architecture on **that** framework with a balanced sampler, the seg results were also pretty good. So the thing is, using your architecture with **their** sampler doesn't work.... Is there anything I missed?

**Comment by YilinLiu97** [comment]
Hi @josedolz , I have another question: do you have the issue that there are always a few subjects that have unusually low accuracy while others are fine, even though all subjects are used for training? Or, some of the classes always have very low dice scores. I suppose that it again has to do with the sampling. 

Thanks.

**Comment by YilinLiu97** [comment]
Hi @josedolz, sorry for bothering you again, but I still can't replicate the result. I found that it worked well only when I added a downsampled pathway just like deepmedic, which makes sense since both context and resolution are maintained. But LiviaNET architecture seems only maintain the resolution and I'm still wondering why it performed so well without a larger context. thanks so much.

**Comment by josedolz** [comment]
Hi @YilinLiu97 

I am sorry to hear that you are still experiencing these problems when trying to replicate LiviaNET in other frameworks. I hope you understand that it is very difficult for me to help you if you work in other framework, as many things are out of control.

If the application requires large receptive fields, having more context, such as dual path-way (DeepMedic), dilation convolutions, or some other strategies actually help. I can tell you for example that for tumour segmentation LiviaNET does not get on top of the methods, particularly because of that (large targets). For subcortical brain segmentation, or brain tissue, for example, there are not requirements to have large receptive fields, that is why we do not need larger context.

**Comment by YilinLiu97** [comment]
Hi @josedolz, thank you for the help! Fortunately I have successfully replicate the results :) 

I just have one more question that really bothers me: there are always 1 or more subjects that have significantly lower scores than others. At first I thought that it was because of the data itself (bad data maybe), but later I found that different framework gave me different 'bad' subjects, i.e., the subject with unreasonably low dice scores on one framework can actually have a much higher dice scores on another framework, or vice versa. Is this common? I've even tried only training on subjects data with particularly low dice scores but still get poor dice overlaps, either on LiviaNET or the other framework.

**Comment by josedolz** [comment]
Hi @YilinLiu97 

I am glad you have finally solved your problem reproducing the results in a different DL framework. If you want to make public that code I can include a link to it at the end of the ReadMe file in this repository, so that people can have the network in different frameworks. Btw, would you mind to explain which was the cause of the differences and how you solved it?

For your question about the different performances on different scans, actually it is weird. In the ABIDE dataset we found poor performance in some scans, but when we looked at them we realized there were bad images (for example with a lot of artifacts and very noisy images. In the Neuroimage paper there are some examples of such scans).

**Comment by YilinLiu97** [comment]
Hi @josedolz , sorry for my late reply and thanks for your suggestion. 

Sure, once everything is fixed I will let you know and make public that code and anything else if necessary. 

The cause seems to be in the pre-processing: I accidentally only normalized the foreground. But I still don't fully understand why it makes that much difference because with another architecture it worked well (even only the foregrounds are normalized).
